pkg/policy/types/entry.go

```
@@ -0,0 +1,245 @@
+}
+
+// DenyEntry returns a MapStateEntry with maximum precedence for a deny entry
+func DenyEntry() MapStateEntry {
+	return MapStateEntry{Precedence: MaxDenyPrecedence}
+}
+
+// WithDeny returns the entry 'e' with the precedence set to deny, or allow without proxy redirect,
+// depending on the value of 'isDeny' parameter
+func (e MapStateEntry) WithDeny(isDeny bool) MapStateEntry {
+	if isDeny {
+		e.Precedence |= PrecedenceDeny | PrecedenceProxyPriorityMask
+	} else {
+		e.Precedence &= ^(PrecedenceDeny | PrecedenceProxyPriorityMask)
+	}
+	return e
+}
+
+// WithProxyPriority returns a MapStateEntry with the given listener priority
+// 0 - default (low) priority for all proxy redirects
+// 1 - highest listener priority
+// ..
+// 100 - lowest (non-default) listener priority
+func (e MapStateEntry) WithProxyPriority(priority uint8) MapStateEntry {
+	if e.ProxyPort != 0 {
+		// Clear the proxy precedence bits
+		e.Precedence &= ^(PrecedenceDeny | PrecedenceProxyPriorityMask)
+		if priority > 0 {
+			if priority > MaxListenerPriority {
+				priority = MaxListenerPriority
+			}
+			// invert the priority to a precedence where higher number has the
+			// precedence, priority 1 becomes 126, 100 -> 27
+			e.Precedence += PrecedenceProxyPriorityMask - Precedence(priority)
+		} else {
+			e.Precedence += 1 // proxy port without explicit priority
+		}
+	}
+	return e
+}
+
+// WithProxyPort return the MapStateEntry with proxy port set at the default precedence
+func (e MapStateEntry) WithProxyPort(proxyPort uint16) MapStateEntry {
+	e.ProxyPort = proxyPort
+	e.Precedence &= ^(PrecedenceDeny | PrecedenceProxyPriorityMask)
+	e.Precedence += 1 // proxy port without explicit priority
+	return e
+}
+
+// Merge is only called for entries whose precedence may differ only for the proxy port priority
+// value.
+func (e *MapStateEntry) Merge(entry MapStateEntry) {
+	if entry.Precedence^e.Precedence >= PrecedenceDeny {
+		panic("Merge called for incompatible entries")
+	}
```

Please no random panics deep in random merge functions. If there's a chance that bad input comes to this function, maybe return an error up the stack to a point where we can make a reasonable choice about how to ignore some portion of policy state preparation, log the error, and gracefully continue under a "fail-safe" mode.

I've discussed `panic` in a bit more detail here: https://github.com/cilium/cilium/issues/33364

---

Removed the panic, was a leftover from maturing this change!